src/diverse_seq/cluster.py

```diff
@@ -123,158 +289,134 @@ def main(self, seq_names: list[str]) -> PhyloNode:
             if self._numprocs != 1
             else nullcontext()
         )
+
+        seq_names = seqs.names
+        seq_arrays = [self._s2a(seqs.get_seq(name)) for name in seq_names]  # pylint: disable=not-callable
+
         with self._progress, self._executor:
             if self._distance_mode == "mash":
-                distances = self.mash_distances(seq_names)
+                distances = self.mash_distances_parallel(seq_arrays)
             elif self._distance_mode == "euclidean":
-                distances = self.euclidean_distances(seq_names)
+                distances = self.euclidean_distances_parallel(seq_arrays)
             else:
                 msg = f"Unexpected distance {self._distance_mode}."
                 raise ValueError(msg)
-            return self.make_cluster_tree(seq_names, distances)
+            return make_cluster_tree(seq_names, distances, progress=self._progress)
 
-    def make_cluster_tree(
-        self,
-        seq_names: Sequence[str],
-        pairwise_distances: numpy.ndarray,
-    ) -> PhyloNode:
-        """Given pairwise distances between sequences, construct a cluster tree.
+    def mash_distances_parallel(self, seq_arrays: Sequence[SeqArray]) -> numpy.ndarray:
```

Comment:
Is there a better kind of name you'd suggest? I'd like to avoid `mash_distances`/`euclidean_distances` as those names are taken by the standard functions. So far, I've just updated the docstrings to say that when one process is specified that it runs in serial. It being parallel capable was the special thing about the function hence the name (and I feel like if the user specifies one process it running in serial is a given).

I guess one option would be to rename the other functions (`mash_distances`/`euclidean_distances`) to `mash_distances_serial` and `euclidean_distances_serial`, and these to `mash_distances`/`euclidean_distances`?

Reply:
As apps, they're intended to be used as callables not as class objects. So I suggest using private method names, then the leading _ helps distinguish them. Also, since the measure is specified to the constructor, you can create another class attribute called `self.._calc_dist` and assign the appropriate method based on the input argument. This will simplify the `main()` method. For example
```python
def ___init___(self, ..., distance, ...)
    self._calc_dist = self._mash_dist if distance == "mash" else ...
```
and modify `main()` accordingly.

Reply:
Ah, that's a great idea